builder/dockerfile/builder.go

```diff
@@ -0,0 +1,291 @@
+package dockerfile
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/builder"
+	"github.com/docker/docker/builder/dockerfile/parser"
+	"github.com/docker/docker/daemon"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/pkg/ulimit"
+	"github.com/docker/docker/runconfig"
+)
+
+var validCommitCommands = map[string]bool{
+	"cmd":        true,
+	"entrypoint": true,
+	"env":        true,
+	"expose":     true,
+	"label":      true,
+	"onbuild":    true,
+	"user":       true,
+	"volume":     true,
+	"workdir":    true,
+}
+
+// BuiltinAllowedBuildArgs is list of built-in allowed build args
+var BuiltinAllowedBuildArgs = map[string]bool{
+	"HTTP_PROXY":  true,
+	"http_proxy":  true,
+	"HTTPS_PROXY": true,
+	"https_proxy": true,
+	"FTP_PROXY":   true,
+	"ftp_proxy":   true,
+	"NO_PROXY":    true,
+	"no_proxy":    true,
+}
+
+// Config constitutes the configuration for a Dockerfile builder.
+type Config struct {
+	// only used if Dockerfile has to be extracted from Context
+	DockerfileName string
+
+	Verbose     bool
+	UseCache    bool
+	Remove      bool
+	ForceRemove bool
+	Pull        bool
+	BuildArgs   map[string]string // build-time args received in build context for expansion/substitution and commands in 'run'.
+
+	// resource constraints
+	// TODO: factor out to be reused with Run ?
+
+	Memory       int64
+	MemorySwap   int64
+	CPUShares    int64
+	CPUPeriod    int64
+	CPUQuota     int64
+	CPUSetCpus   string
+	CPUSetMems   string
+	CgroupParent string
+	Ulimits      []*ulimit.Ulimit
+}
+
+// Builder is a Dockerfile builder
+// It implements the builder.Builder interface.
+type Builder struct {
+	*Config
+
+	Stdout io.Writer
+	Stderr io.Writer
+
+	docker  builder.Docker
+	context builder.Context
+
+	dockerfile       *parser.Node
+	runConfig        *runconfig.Config // runconfig for cmd, run, entrypoint etc.
+	flags            *BFlags
+	tmpContainers    map[string]struct{}
+	image            string // imageID
+	noBaseImage      bool
+	maintainer       string
+	cmdSet           bool
+	disableCommit    bool
+	cacheBusted      bool
+	cancelled        chan struct{}
+	cancelOnce       sync.Once
+	allowedBuildArgs map[string]bool // list of build-time args that are allowed for expansion/substitution and passing to commands in 'run'.
+
+	// TODO: remove once docker.Commit can receive a tag
+	id           string
+	activeImages []string
+}
+
+// NewBuilder creates a new Dockerfile builder from an optional dockerfile and a Config.
+// If dockerfile is nil, the Dockerfile specified by Config.DockerfileName,
+// will be read from the Context passed to Build().
+func NewBuilder(dockerfile io.ReadCloser, config *Config) (b *Builder, err error) {
```

Comment:
Won't be the case for client side builder. This is so that we can provide a dockerfile independent from the context. It will get the Dockerfile from the context only if this dockerfile is nil. Which is always the case for the server side implementation.